### PR TITLE
added image processing extras

### PR DESCRIPTION
**Pull request by likangning93**
In anticipation of ambient occlusion and other future image processing stages, I added extras for use with jimp (https://www.npmjs.com/package/jimp) in the form of a gltf-level scratch `Jimp` image and `Jimp` copies of all the existing images, stored as `jimpImage` in `extras._pipeline.`

Since this won't be used for every pipeline configuration, I tried to make it as optional as possible.

I also added relevant specs.

### CODE REVIEW

**Comment by lilleyse** [comment]
I'm a bit torn, because while it's faster to make the image processing an optional step, it makes the code a bit more confusing. It would be kind of nice if image processing was completely hidden from `gltfPipeline.js`, and instead it always decoded in `loadGltfUris` and encodes in `writeSource` (but only if the data was changed, this can be marked by a flag in the image's `extras` that is set by the AO stage/atlas stage etc).

An alternative is to continue to allow `readGltf` and `loadGltfUris` to take the `imageProcess` flag, but not have an `encodeImages` stage in `gltfPipeline`, instead always encode in `writeSource` using the flag idea above.

Let me know what you think, or if I'm overlooking something.

**Comment by pjcozzi** [comment]
Can't this be handled with each stage declaring its dependencies?  See #75.

**Comment by likangning93** [comment]
I really like option 2 for re-encoding. I also toyed with the idea of
changing 'loadGltfUris' to take an options parameter instead of 'basePath',
which might let us hide the image processing more. I'll give it a try.

**Comment by likangning93** [comment]
Updated! As discussed offline, we decided to keep the `encodeImages` stage where it is for now. It seems like `writeSource` (which handles moving extras back into the json) is currently tied to `writeGltf` and not to `writeBinaryGltf`, and we may want to move it to come before the actual file writing options. When that happens we will move `encodeImages` over to that. `encodeImages` also now checks for an `imageChanged` flag before encoding.

I also updated `loadGltfUris` to take in an options parameter instead of just `basePath` and updated all relevant specs accordingly.

**Comment by likangning93** [comment]
merged in master.

File: lib/encodeImages.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
+'use strict';
+var Jimp = require('jimp');
+
+module.exports = encodeImages;
+var async = require('async');
```
Comment (lilleyse) [inline lib/encodeImages.js]: Fix the ordering here.

File: lib/encodeImages.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+function encodeImages(gltfWithExtras, callback) {
```
Comment (lilleyse) [inline lib/encodeImages.js]: Change `gltfWithExtras` to `gltf` to follow the other stages.

File: lib/encodeImages.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            if (callback) {
+                callback(err);
+            }
+            else{
+                throw err;
+            }
+        }
+        else if (callback) {
+            callback(gltfWithExtras);
+        }
```
Comment (lilleyse) [inline lib/encodeImages.js]: Right now the callback to `encodeImage` expects a gltf object, not an error. Here you can just simplify to

```javascript
if (err) {
    throw err;
}
callback(gltf)
```

File: lib/loadGltfUris.js [hunk 4]
```diff
@@ -8,18 +8,37 @@ var DeveloperError = Cesium.DeveloperError;
...
+                }
+                gltf.extras._pipeline.jimpScratch = image;
+                callback(err, gltf);
+            });
+        };
+    }
+
     async.each(['buffers', 'images', 'shaders'], function(name, asyncCallback) {
-        loadURI(gltf, basePath, name, asyncCallback);
+        loadURI(gltf, basePath, name, asyncCallback, imageProcess);
```
Comment (lilleyse) [inline lib/loadGltfUris.js]: Usually the callback should be the last argument.

File: lib/readGltf.js [hunk 5]
```diff
@@ -10,14 +10,13 @@ var DeveloperError = Cesium.DeveloperError;
 
 module.exports = readGltf;
 
-function readGltf(gltfPath, callback) {
+function readGltf(gltfPath, options, callback, imageProcess) {
```
Comment (lilleyse) [inline lib/readGltf.js]: You don't use `imageProcess` anymore right?

File: lib/readGltf.js [hunk 6]
```diff
@@ -35,11 +34,11 @@ function readGltf(gltfPath, callback) {
...
             addPipelineExtras(gltf);
         }
-        loadGltfUris(gltf, filePath, function(err, gltf) {
+        loadGltfUris(gltf, options, function(err, gltf) {
             if (err) {
                 throw err;
             }
             callback(gltf);
-        });
+        }, imageProcess);
```
Comment (lilleyse) [inline lib/readGltf.js]: Same comment here.

File: lib/readGltf.js [hunk 7]
```diff
@@ -10,14 +10,13 @@ var DeveloperError = Cesium.DeveloperError;
...
 module.exports = readGltf;
 
-function readGltf(gltfPath, callback) {
+function readGltf(gltfPath, options, callback, imageProcess) {
     if (!defined(gltfPath)) {
         throw new DeveloperError('Input path is undefined.');
     }
     
     var fileExtension = path.extname(gltfPath);
-    var filePath = path.dirname(gltfPath);
```
Comment (lilleyse) [inline lib/readGltf.js]: I know this change is now everywhere, but I think `options.basePath` should just be an optional argument for `readGltf`. Almost all of the time the `basePath` can be computed like this.

File: specs/lib/encodeImagesSpec.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+var encodeImages = require('../../lib/encodeImages');
+var loadGltfUris = require('../../lib/loadGltfUris');
+var dataUriToBuffer = require('data-uri-to-buffer');
+var imageUri = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAICAIAAABLbSncAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsMAAA7DAcdvqGQAAADTSURBVBhXAcgAN/8B49/cCAQAyukB2vAB+vz/Ig7+QR0B+PwAAezj3LDfAqnZ/wMB/xwPBwUEAiMK91Uj/gLN6wGj1fs9IyEwGxoUCPotFQC75g7rASECvd/6KxwkVCHFUiTX9P4Xq9L8ZjUD+vWyAaHK+CUX+pujFBYRH1NR2vUAANrLX8zXyAJCHOWnsj3d7frR4+XLymz24Y6ZuKI7LGUE/vcBEAglAQTR/P/9nbmV8fUAYURiTjRqAeXi6AgFDMDWptPiwP///isdPEIsXvr8+Tj0Y5s8qCp8AAAAAElFTkSuQmCC';
+var basePath = './specs/data/boxTexturedUnoptimized/';
+var clone = require('clone');
+var Jimp = require('jimp');
+
+describe('encodeImages', function() {
+    var imageData;
```
Comment (lilleyse) [inline specs/lib/encodeImagesSpec.js]: This var isn't used.

File: specs/lib/encodeImagesSpec.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        loadGltfUris(gltfClone, options, function() {
+            var pipelineExtras0001 = gltfClone.images.Image0001.extras._pipeline;
+            var pipelineExtras0002 = gltfClone.images.Image0002.extras._pipeline;
+
+            encodeImages(gltfClone, function() {
+                expect(bufferEqual(pipelineExtras0001.source, imageBuffer)).toBe(true);
+                expect(bufferEqual(pipelineExtras0002.source, imageBuffer)).toBe(true);
+                done();
+            });
+        }, true);
```
Comment (lilleyse) [inline specs/lib/encodeImagesSpec.js]: Remove this last parameter.

File: specs/lib/encodeImagesSpec.js [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+                    expect(image.bitmap.height).toEqual(10);
+                    expect(jimpImage001.bitmap.height).toEqual(10);
+                    Jimp.read(pipelineExtras0002.source, function(jimpErr, image) {
+                        expect(image.bitmap.height).toEqual(8);
+                        expect(jimpImage002.bitmap.height).toEqual(8);
+                        done();
+                    });
+                });
+            });
+        }, true);
```
Comment (lilleyse) [inline specs/lib/encodeImagesSpec.js]: And this.

File: specs/lib/gltfPipelineSpec.js [hunk 11]
```diff
@@ -91,7 +101,10 @@ describe('gltfPipeline', function() {
         var spy = spyOn(fs, 'writeFile').and.callFake(function(file, data, callback) {
             callback();
         });
-        var options = { 'binary' : true };
+        var options = {
+            basePath: basePath,
+            'binary' : true
```
Comment (lilleyse) [inline specs/lib/gltfPipelineSpec.js]: Remove quotation marks.

File: lib/readGltf.js (hunk 5 above)
Comment (likangning93) [inline lib/readGltf.js]: It's still needed to signal that a `jimp` needs to get added.

Comment (lilleyse) [inline lib/readGltf.js]: I mean, isn't it now `options.imageProcess`?

Comment (likangning93) [inline lib/readGltf.js]: I kind of unpacked the options further up, should I just be doing `options.xxx`?

Comment (likangning93) [inline lib/readGltf.js]: Ah. I was looking at the wrong file.

**Comment by likangning93** [comment]
Updated!

**Comment by likangning93** [comment]
@pjcozzi integrating image processing into `runStages` might be a little messy since the encoding and decoding have to happen at the beginning and end of the pipeline and it seems like `runStages` handles the stages that are more synchronous.

I imagine we could add something before runStages that checks if the pipeline needs image processing extras, but that could get messy if we have to add anything else that wraps the pipeline like this.

File: specs/lib/cacheOptimizeIndicesSpec.js [hunk 12]
```diff
@@ -5,6 +5,7 @@ var readGltf = require('../../lib/readGltf');
 var readAccessor = require('../../lib/readAccessor');
 var writeAccessor = require('../../lib/writeAccessor');
 
+var basePath = './specs/data/boxTexturedUnoptimized/';
```
Comment (lilleyse) [inline specs/lib/cacheOptimizeIndicesSpec.js]: There are a few lines like this that aren't used. Otherwise this PR looks ready to me.

**Comment by likangning93** [comment]
ok. Updated!

**Comment by lasalvavida** [comment]
> @pjcozzi integrating image processing into runStages might be a little messy since the encoding and decoding have to happen at the beginning and end of the pipeline and it seems like runStages handles the stages that are more synchronous.

We should adopt a promise style architecture for both synchronous and asynchronous stages; it would allow all stages to be run in the same fashion.

**Comment by pjcozzi** [comment]
> We should adopt a promise style architecture for both synchronous and asynchronous stages; it would allow all stages to be run in the same fashion.

Yes, please add it to #1.

File: lib/gltfPipeline.js [hunk 13]
```diff
@@ -54,11 +55,16 @@ function processJSONWithExtras(gltfWithExtras, options, callback) {
     if (options.quantize) {
         quantizeAttributes(gltfWithExtras);
     }
-    callback(gltfWithExtras);
+    if (options.imageProcess) {
+        encodeImages(gltfWithExtras, callback);
+    }
```
Comment (pjcozzi) [inline lib/gltfPipeline.js]: `} else {`

File: package.json [hunk 14]
```diff
@@ -38,6 +38,7 @@
     "data-uri-to-buffer": "0.0.4",
     "datauri": "^1.0.0-alpha.5",
     "image-size": "^0.5.0",
+    "jimp": "^0.2.24",
```
Comment (pjcozzi) [inline package.json]: Add this to the third-party section of [LICENSE.md](https://github.com/AnalyticalGraphicsInc/gltf-pipeline/blob/master/LICENSE.md).  In general, can we please update that section to match the libraries here in `dependencies`?

**Pull request merged**